File: generated_code/py/unittest_2.py
import unittest
from typing import Optional, List

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

# Implementation placeholder for testing
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        dummy = ListNode(0)
        curr = dummy
        carry = 0
        while l1 or l2 or carry:
            val1 = l1.val if l1 else 0
            val2 = l2.val if l2 else 0
            
            total = val1 + val2 + carry
            carry = total // 10
            curr.next = ListNode(total % 10)
            
            curr = curr.next
            if l1: l1 = l1.next
            if l2: l2 = l2.next
        return dummy.next

class TestAddTwoNumbers(unittest.TestCase):
    
    # Helper methods to convert between python lists and linked lists
    def to_linked_list(self, arr: List[int]) -> Optional[ListNode]:
        if not arr:
            return None
        dummy = ListNode(0)
        curr = dummy
        for val in arr:
            curr.next = ListNode(val)
            curr = curr.next
        return dummy.next

    def to_python_list(self, head: Optional[ListNode]) -> List[int]:
        arr = []
        curr = head
        while curr:
            arr.append(curr.val)
            curr = curr.next
        return arr

    def assertLinkedListEqual(self, l1: Optional[ListNode], l2_arr: List[int]):
        self.assertEqual(self.to_python_list(l1), l2_arr)

    # --- Example Test Cases ---
    
    def test_example_1(self):
        """Test with l1 = [2,4,3] and l2 = [5,6,4] -> [7,0,8]"""
        l1 = self.to_linked_list([2, 4, 3])
        l2 = self.to_linked_list([5, 6, 4])
        sol = Solution()
        result = sol.addTwoNumbers(l1, l2)
        self.assertLinkedListEqual(result, [7, 0, 8])

    def test_example_2(self):
        """Test with l1 = [0] and l2 = [0] -> [0]"""
        l1 = self.to_linked_list([0])
        l2 = self.to_linked_list([0])
        sol = Solution()
        result = sol.addTwoNumbers(l1, l2)
        self.assertLinkedListEqual(result, [0])

    def test_example_3(self):
        """Test with different lengths and multiple carries: l1 = [9,9,9,9,9,9,9], l2 = [9,9,9,9]"""
        l1 = self.to_linked_list([9, 9, 9, 9, 9, 9, 9])
        l2 = self.to_linked_list([9, 9, 9, 9])
        sol = Solution()
        result = sol.addTwoNumbers(l1, l2)
        self.assertLinkedListEqual(result, [8, 9, 9, 9, 0, 0, 0, 1])

    # --- Edge Cases & Constraint Tests ---

    def test_single_digit_with_carry(self):
        """Test single digit addition that results in a carry: 5 + 5 = 10 -> [0, 1]"""
        l1 = self.to_linked_list([5])
        l2 = self.to_linked_list([5])
        sol = Solution()
        result = sol.addTwoNumbers(l1, l2)
        self.assertLinkedListEqual(result, [0, 1])

    def test_one_list_longer_no_carry(self):
        """Test where one list is longer and no carry transfers to the extended digits"""
        l1 = self.to_linked_list([1, 2])
        l2 = self.to_linked_list([3, 4, 5, 6])
        sol = Solution()
        result = sol.addTwoNumbers(l1, l2)
        self.assertLinkedListEqual(result, [4, 6, 5, 6])

    def test_one_list_longer_with_carry_propagation(self):
        """Test where carry propagates into the remaining digits of the longer list"""
        l1 = self.to_linked_list([9, 9])
        l2 = self.to_linked_list([1, 0, 0, 4])
        sol = Solution()
        result = sol.addTwoNumbers(l1, l2)
        self.assertLinkedListEqual(result, [0, 0, 1, 4])

    def test_max_length_constraint(self):
        """Test maximum constraint boundary (100 nodes each) filled with max digits"""
        l1 = self.to_linked_list([9] * 100)
        l2 = self.to_linked_list([9] * 100)
        sol = Solution()
        result = sol.addTwoNumbers(l1, l2)
        
        # 99...99 + 99...99 = 199...98
        expected = [8] + ([9] * 99) + [1]
        self.assertLinkedListEqual(result, expected)

if __name__ == '__main__':
    unittest.main()

File: generated_code/cpp/unittest_2.cpp
#include <gtest/gtest.h>
#include <vector>

// Definition for singly-linked list provided by the problem
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

// Solution class to be tested
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode dummy(0);
        ListNode* tail = &dummy;
        int carry = 0;

        while (l1 != nullptr || l2 != nullptr || carry != 0) {
            int digit1 = (l1 != nullptr) ? l1->val : 0;
            int digit2 = (l2 != nullptr) ? l2->val : 0;

            int sum = digit1 + digit2 + carry;
            carry = sum / 10;
            tail->next = new ListNode(sum % 10);
            tail = tail->next;

            if (l1 != nullptr) l1 = l1->next;
            if (l2 != nullptr) l2 = l2->next;
        }

        return dummy.next;
    }
};

// Helper function to create a linked list from a vector
ListNode* createLinkedList(const std::vector<int>& values) {
    if (values.empty()) return nullptr;
    ListNode* head = new ListNode(values[0]);
    ListNode* current = head;
    for (size_t i = 1; i < values.size(); ++i) {
        current->next = new ListNode(values[i]);
        current = current->next;
    }
    return head;
}

// Helper function to convert a linked list back to a vector for verification
std::vector<int> linkedListToVector(ListNode* head) {
    std::vector<int> result;
    while (head != nullptr) {
        result.push_back(head->val);
        head = head->next;
    }
    return result;
}

// Helper function to free allocated linked list memory
void freeLinkedList(ListNode* head) {
    while (head != nullptr) {
        ListNode* temp = head;
        head = head->next;
        delete temp;
    }
}

class AddTwoNumbersTest : public ::testing::Test {
protected:
    Solution solution;
};

// Test Case 1: Example 1 from problem description
TEST_F(AddTwoNumbersTest, TestExample1) {
    ListNode* l1 = createLinkedList({2, 4, 3});
    ListNode* l2 = createLinkedList({5, 6, 4});
    
    ListNode* result = solution.addTwoNumbers(l1, l2);
    std::vector<int> expected = {7, 0, 8};
    
    EXPECT_EQ(linkedListToVector(result), expected);
    
    freeLinkedList(l1);
    freeLinkedList(l2);
    freeLinkedList(result);
}

// Test Case 2: Example 2 from problem description (Single elements, zeros)
TEST_F(AddTwoNumbersTest, TestExample2) {
    ListNode* l1 = createLinkedList({0});
    ListNode* l2 = createLinkedList({0});
    
    ListNode* result = solution.addTwoNumbers(l1, l2);
    std::vector<int> expected = {0};
    
    EXPECT_EQ(linkedListToVector(result), expected);
    
    freeLinkedList(l1);
    freeLinkedList(l2);
    freeLinkedList(result);
}

// Test Case 3: Example 3 from problem description (Different lengths, multiple carries)
TEST_F(AddTwoNumbersTest, TestExample3) {
    ListNode* l1 = createLinkedList({9, 9, 9, 9, 9, 9, 9});
    ListNode* l2 = createLinkedList({9, 9, 9, 9});
    
    ListNode* result = solution.addTwoNumbers(l1, l2);
    std::vector<int> expected = {8, 9, 9, 9, 0, 0, 0, 1};
    
    EXPECT_EQ(linkedListToVector(result), expected);
    
    freeLinkedList(l1);
    freeLinkedList(l2);
    freeLinkedList(result);
}

// Test Case 4: Edge Case - One list is shorter than the other (L1 < L2)
TEST_F(AddTwoNumbersTest, TestShorterFirstList) {
    ListNode* l1 = createLinkedList({1, 2});
    ListNode* l2 = createLinkedList({9, 7, 9});
    
    ListNode* result = solution.addTwoNumbers(l1, l2);
    std::vector<int> expected = {0, 0, 0, 1}; // 21 + 979 = 1000 -> [0, 0, 0, 1]
    
    EXPECT_EQ(linkedListToVector(result), expected);
    
    freeLinkedList(l1);
    freeLinkedList(l2);
    freeLinkedList(result);
}

// Test Case 5: Edge Case - Minimum node constraints (1 node each, non-zero values)
TEST_F(AddTwoNumbersTest, TestMinimumNodeConstraint) {
    ListNode* l1 = createLinkedList({5});
    ListNode* l2 = createLinkedList({5});
    
    ListNode* result = solution.addTwoNumbers(l1, l2);
    std::vector<int> expected = {0, 1};
    
    EXPECT_EQ(linkedListToVector(result), expected);
    
    freeLinkedList(l1);
    freeLinkedList(l2);
    freeLinkedList(result);
}

// Test Case 6: Edge Case - Final addition creates an extra carry node
TEST_F(AddTwoNumbersTest, TestFinalCarryCreation) {
    ListNode* l1 = createLinkedList({1});
    ListNode* l2 = createLinkedList({9, 9});
    
    ListNode* result = solution.addTwoNumbers(l1, l2);
    std::vector<int> expected = {0, 0, 1}; // 1 + 99 = 100 -> [0, 0, 1]
    
    EXPECT_EQ(linkedListToVector(result), expected);
    
    freeLinkedList(l1);
    freeLinkedList(l2);
    freeLinkedList(result);
}

// Test Case 7: Edge Case - High capacity input representation (Maximum constraints up to 100 nodes simulated)
TEST_F(AddTwoNumbersTest, TestLargeValueConstraint) {
    std::vector<int> largeNum1(100, 9); // 100 nodes of 9s
    std::vector<int> largeNum2(100, 9); // 100 nodes of 9s
    
    ListNode* l1 = createLinkedList(largeNum1);
    ListNode* l2 = createLinkedList(largeNum2);
    
    ListNode* result = solution.addTwoNumbers(l1, l2);
    
    // 99...99 + 99...99 results in 8 followed by ninety-nine 9s followed by 1
    std::vector<int> expected(101, 9);
    expected[0] = 8;
    expected[100] = 1;
    
    EXPECT_EQ(linkedListToVector(result), expected);
    
    freeLinkedList(l1);
    freeLinkedList(l2);
    freeLinkedList(result);
}

int main(int argc, char **argv) {
    ::testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}